server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -306,6 +352,73 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongOrDateSort(SearchContext searchContext, IndexReader reader) throws IOException {
```

Comment:
We should not activate the optim if trackTotalHits is true or set to Integer.MAX_VALUE since we cannot skip hits in this case. I think that moving these checks to `TopDocsCollectorContext#SimpleTopDocsCollectorContext` would be easier since you can find the final value of `trackTotalHits` easily. We should also not activate this optim if aggregations are part of the query because we don't apply max-score in this case either.